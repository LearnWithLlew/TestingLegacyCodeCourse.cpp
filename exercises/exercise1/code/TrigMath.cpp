#include "TrigMath.h"

double TrigMath::SinRaw(int idx) {
    return SIN_TABLE[idx & SIN_MASK];
}

double TrigMath::CosRaw(int idx) {
    return SIN_TABLE[(idx + COS_OFFSET) & SIN_MASK];
}

double TrigMath::Mxatan(double arg) {
    auto argsq = arg * arg;
    auto value = (((p4 * argsq + p3) * argsq + p2) * argsq + p1) * argsq + p0;
    value /= ((((argsq + q4) * argsq + q3) * argsq + q2) * argsq + q1) * argsq + q0;
    return value * arg;
}

double TrigMath::Msatan(double arg) {
    if (arg < sq2m1)
        return Mxatan(arg);
    if (arg > sq2p1)
        return HALF_PI - Mxatan(1 / arg);
    return HALF_PI / 2 + Mxatan((arg - 1) / (arg + 1));
}

void TrigMath::CreateSinTable() {
    for (auto i = 0; i < SIN_SIZE; i++)
        SIN_TABLE[i] = sin(i * TWO_PI / SIN_SIZE);
}

TrigMath::TrigMath() {
    CreateSinTable();
}

double TrigMath::Sin(double angle) {
    return SinRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

int TrigMath::Floor(double a) {
    return (int) a;
}

double TrigMath::Cos(double angle) {
    return CosRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

double TrigMath::Tan(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return SinRaw(idx) / CosRaw(idx);
}

double TrigMath::Sec(double angle) {
    return 1 / Cos(angle);
}

double TrigMath::Cot(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return CosRaw(idx) / SinRaw(idx);
}

double TrigMath::Asin(double value) {
    if (value > 1)
        return NAN;
    if (value < 0)
        return -Asin(-value);
    auto temp = sqrt(1 - value * value);
    if (value > 0.7)
        return HALF_PI - Msatan(temp / value);
    return Msatan(value / temp);
}

double TrigMath::Acos(double value) {
    if (value > 1 || value < -1)
        return NAN;
    return HALF_PI - Asin(value);
}

double TrigMath::Atan(double value) {
    if (value > 0)
        return Msatan(value);
    return -Msatan(-value);
}

double TrigMath::Atan2(double y, double x) {
    if (y + x == y)
        return y >= 0 ? HALF_PI : -HALF_PI;
    y = Atan(y / x);
    if (x < 0) {
        if (y <= 0) {
            return y + PI;
        } else {
            return y - PI;
        }
    }
    return y;
}

double TrigMath::Acsc(double value) {
    if (value == 0)
        return NAN;
    return Asin(1 / value);
}

double TrigMath::Asec(double value) {
    if (value == 0)
        return NAN;
    return Acos(1 / value);
}

double TrigMath::Acot(double value) {
    if (value == 0)
        return NAN;
    if (value > 0)
        return Atan(1 / value);
    return Atan(1 / value) + PI;
}

