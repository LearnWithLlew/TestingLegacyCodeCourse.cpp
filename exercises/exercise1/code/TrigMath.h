//
// Created by LLEWELLYN FALCO on 12/10/17.
//

#ifndef CATCHPLAYGROUND_TRIGMATH_H
#define CATCHPLAYGROUND_TRIGMATH_H

#define _USE_MATH_DEFINES
#include <math.h>
#include <cmath>

class TrigMath {
public:
    const double PI = M_PI;
    const double SQUARED_PI = PI * PI;
    const double HALF_PI = PI / 2;
    const double QUARTER_PI = HALF_PI / 2;
    const double TWO_PI = 2 * PI;
    const double THREE_PI_HALVES = TWO_PI - HALF_PI;
    const double DEG_TO_RAD = PI / 180;
    const double HALF_DEG_TO_RAD = PI / 360;
    const double RAD_TO_DEG = 180 / PI;
    double SQRT_OF_TWO = std::sqrt(2);
    double HALF_SQRT_OF_TWO = SQRT_OF_TWO / 2;
private  :

    const int SIN_SIZE = 100000;
    double SIN_TABLE[100000];
    const int SIN_MASK = SIN_SIZE - 1;
    const double SIN_CONVERSION_FACTOR = SIN_SIZE / TWO_PI;

    const int COS_OFFSET = SIN_SIZE / 4;

//Arc trig
    const double sq2p1 = 2.414213562373095048802;

    const double sq2m1 = 0.414213562373095048802;
    const double p4 = 0.161536412982230228262E2;
    const double p3 = 0.26842548195503973794141E3;
    const double p2 = 0.11530293515404850115428136E4;
    const double p1 = 0.178040631643319697105464587E4;
    const double p0 = 0.89678597403663861959987488E3;
    const double q4 = 0.5895697050844462222791E2;
    const double q3 = 0.536265374031215315104235E3;
    const double q2 = 0.16667838148816337184521798E4;
    const double q1 = 0.207933497444540981287275926E4;
    const double q0 = 0.89678597403663861962481162E3;

    double sinRaw(int idx);

    double cosRaw(int idx);

    double mxatan(double arg);

    double msatan(double arg);

    void createSinTable();

public:

    TrigMath();


    double sin(double angle);

    int floor(double a);


    double cos(double angle);

    double tan(double angle);

    double csc(double angle) {
        return 1 / sin(angle);
    }

    double sec(double angle);


    double cot(double angle);

    double asin(double value);

    double acos(double value);

    double atan(double value);

    double atan2(double y, double x);

    double acsc(double value);

    double asec(double value);

    double acot(double value);


};

#endif //CATCHPLAYGROUND_TRIGMATH_H
