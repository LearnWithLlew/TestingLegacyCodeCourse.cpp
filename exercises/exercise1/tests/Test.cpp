#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"
#include "TrigMath.h"

using namespace ApprovalTests;



TEST_CASE("Test Sin") {
    TrigMath math;
    Approvals::verify(math.Sin(3.4));
}
TEST_CASE("Test Cos") {
    TrigMath math;
    Approvals::verify(math.Cos(3.4));
}
std::string callEverything(double value) {
    std::stringstream s;
    TrigMath math;
    s << "Acos (" << value << ") = " << math.Acos(value) << "\n";
    s << "Acot  (" << value << ") = " <<  math.Acot(value) << "\n";
    s << "Acsc (" << value << ") = " <<  math.Acsc(value) << "\n";
    s << "Asec (" << value << ") = " <<  math.Asec(value) << "\n";
    s << "Asin (" << value << ") = " <<  math.Asin(value) << "\n";
    s << "Atan (" << value << ") = " <<  math.Atan(value) << "\n";
    s << "Atan2 (" << value << ",0.4) = " <<  math.Atan2(value,0.4) << "\n";
    s << "Cos (" << value << ") = "<< math.Cos(value) << "\n";
    s << "Cot (" << value << ") = " <<  math.Cot(value) << "\n";
    s << "Csc (" << value << ") = " <<  math.Csc(value) << "\n";
    s << "Floor (" << value << ") = " <<  math.Floor(value) << "\n";
    s << "Sec (" << value << ") = " <<  math.Sec(value) << "\n";
    s << "Sin  (" << value << ") = " <<  math.Sin(value) << "\n";
    s << "Tan  (" << value << ") = " <<  math.Tan(value) << "\n";


    return s.str();

}
TEST_CASE("Test Everything") {

    Approvals::verify(callEverything(1.0));
}
TEST_CASE("Test Everything 3.14") {

    Approvals::verify(callEverything(3.14));
}
TEST_CASE("Test Everything For both values") {
    std::vector < double> values = { 1.0, 3.14 };
    Approvals::verifyAll("TITLE", values, [](auto input, auto& stream) {
        stream << callEverything(input);
        });
}
TEST_CASE("Test Everything With Combinations") {
    std::vector < double> inputs1 = { -0.1,0.0, 1.0, 3.14 };
    CombinationApprovals::verifyAllCombinations(
        [](auto input1) { return callEverything(input1); }, inputs1);

   
}
TEST_CASE("Test Everything With Combinations for Atan2") {
    std::vector < double> inputs1 = {-0.1,0.0, 1.0, 3.14 };

    CombinationApprovals::verifyAllCombinations(
        [](auto input1, auto input2) { return TrigMath ().Atan2(input1, input2); }, inputs1, inputs1);

   
}