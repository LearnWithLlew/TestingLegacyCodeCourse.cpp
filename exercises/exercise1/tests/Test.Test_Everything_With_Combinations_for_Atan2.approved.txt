(-0.1, -0.1) => -2.35619
(-0.1, 0) => -1.5708
(-0.1, 1) => -0.0996687
(-0.1, 3.14) => -0.0318364
(0, -0.1) => 3.14159
(0, 0) => 1.5708
(0, 1) => 0
(0, 3.14) => 0
(1, -0.1) => 1.67046
(1, 0) => 1.5708
(1, 1) => 0.785398
(1, 3.14) => 0.308316
(3.14, -0.1) => 1.60263
(3.14, 0) => 1.5708
(3.14, 1) => 1.26248
(3.14, 3.14) => 0.785398

